test/ng/rootScopeSpec.js

```diff
@@ -2316,6 +2316,97 @@ describe('Scope', function() {
         }));
 
 
+        // See issue https://github.com/angular/angular.js/issues/16135
+        it('should deallocate the listener array entry', inject(function($rootScope) {
+          var remove1 = $rootScope.$on('abc', noop);
+          $rootScope.$on('abc', noop);
+
+          expect($rootScope.$$listeners['abc'].length).toBe(2);
+
+          remove1();
+
+          expect($rootScope.$$listeners['abc'].length).toBe(1);
+        }));
+
+
+        it('should call next listener when removing current', inject(function($rootScope) {
+          var listener1 = jasmine.createSpy().and.callFake(function() { remove1(); });
```

Comment:
Nit: Giving spies a name makes it a little easier to identify the error (in case the test breaks).

Reply:
Note: We are relying upon scope hoisting of the `var remove1` here.

Reply:
If we felt that is bad, one way around this could be:

```
var listener1 = jasmine.createSpy();
var remove1 = $rootScope.$on('abc', listener1);
listener1.and.callFake(function() { remove1(); };
```